Feat/reward summarization example

TristanThrush opened the pull request:
This PR adds a summarization reward-model tuning example, following the "Learning to Summarize from Human Feedback" [paper](https://arxiv.org/abs/2009.01325).

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_115). All of your documentation changes will be reflected on that endpoint.

natolambert commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,182 @@
+
+    return new_examples
+
+num_proc = 100 # Can adjust to be higher if you have more processors.
```
Is this for different distributed training architectures?

TristanThrush replied:
no it's just an argument that you can add in the datasets map function https://huggingface.co/docs/datasets/process

It should work even if your system doesn't have 100 cpus

natolambert replied:
Cool, maybe add a comment saying that (good plug to HF tools)

TristanThrush replied:
done!

natolambert reviewed (commented):
Added some small comments (a somewhat quick read through, I can re-visit):
* Did you train a model, is it on the hub?
* You mentioned the results match the original paper, can you add that info to the PR?
* Can you pull main and run the linting on examples? `Make style`

TristanThrush commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,182 @@
+        )
+        hidden_states_j = transformer_outputs_j[0][:,-1,:]  # Get the last hidden state
+        rewards_j = self.v_head(hidden_states_j)
+        if input_ids_k is not None:
```
Already in the comment at the top of the class. Can add above this line though

natolambert replied:
Yeah, a lesson from diffusers / open source: repeat yourself, makes people less confused.

TristanThrush replied:
took the class out entirely, per leandro's advice

TristanThrush commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,182 @@
+# Load the human comparisons dataset for tuning the reward model.
+ds = load_dataset("openai/summarize_from_feedback", name="comparisons")
+
+# Define the reward model. Code here is loosely inspired by https://wandb.ai/carperai/summarize_RLHF/reports/Implementing-RLHF-Learning-to-Summarize-with-trlX--VmlldzozMzAwODM2.
```
Yes I can write a docstring! What variables are you referring to.

TristanThrush replied:
I actually took this whole class out for my response to leandro's comment

lvwerra reviewed (commented):
Hi @TristanThrush, this is great - thanks for adding! I looked at the code and was thinking about ways to simplify it a bit further. Here's my proposal:

For the reward model we can use the vanilla `AutoModelForSequenceClassification`. This is also just a linear layer on top of the hidden states. Instead of having a custom model we could then just write a `compute_loss` for the `Trainer` (see `CustomTrainer` example [here](https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/trainer#trainer)).

The advantage of this: 1) simplifies the code a bit and 2) we don't need a custom model class. For 2) this also allows you to use all the `.push_to_hub` and `from_pretrained` functionality to share the reward model without any changes.

If we are already having a custom `compute_loss` function we could just tokenize the dataset beforehand and pad to the max. Inside `compute_loss` we can remove the excessive padding from the batch (I think this could be a one-liner) before doing the forward pass. Then we wouldn't need a custom collator neither and the tokenization could be integrated into the `turn_into_text_classification_format`. 

Regarding the config: since we are not explaining exactly how to use it we could omit it here and refer to the DS integration of the Trainer?

What do you think?

TristanThrush commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,182 @@
+            else self.config.n_embd
+        )
+        self.transformer = model.transformer
+        self.v_head = nn.Linear(self.config.n_embd, 1)
```
There are other reasonable reward models that we could imagine. Like, ones that return a softmax. Not sure if we should write this.

TristanThrush replied:
I actually took this class out entirely

TristanThrush commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -40,13 +40,14 @@ def __init__(self, config):
         model = AutoModelForCausalLM.from_pretrained(config)
         self.config = model.config
-        # gpt-neo models have hidden_size instead of n_embd.
+        # Optional internal config change: gpt-neo models use hidden_size instead of n_embd.
```
What is an "optional internal config change"?

TristanThrush replied:
(actually this part of the code was removed anyway)

TristanThrush commented:
> Hi @TristanThrush, this is great - thanks for adding! I looked at the code and was thinking about ways to simplify it a bit further. Here's my proposal:
> 
> For the reward model we can use the vanilla `AutoModelForSequenceClassification`. This is also just a linear layer on top of the hidden states. Instead of having a custom model we could then just write a `compute_loss` for the `Trainer` (see `CustomTrainer` example [here](https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/trainer#trainer)).
> 
> The advantage of this: 1) simplifies the code a bit and 2) we don't need a custom model class. For 2) this also allows you to use all the `.push_to_hub` and `from_pretrained` functionality to share the reward model without any changes.
> 
> If we are already having a custom `compute_loss` function we could just tokenize the dataset beforehand and pad to the max. Inside `compute_loss` we can remove the excessive padding from the batch (I think this could be a one-liner) before doing the forward pass. Then we wouldn't need a custom collator neither and the tokenization could be integrated into the `turn_into_text_classification_format`.
> 
> Regarding the config: since we are not explaining exactly how to use it we could omit it here and refer to the DS integration of the Trainer?
> 
> What do you think?

Thanks leandro, makes sense! I'm getting roped into the rlhf mturk stuff right now but I will try to address later in the week.

TristanThrush commented:
> Added some small comments (a somewhat quick read through, I can re-visit):
> 
> * Did you train a model, is it on the hub?
> * You mentioned the results match the original paper, can you add that info to the PR?
> * Can you pull main and run the linting on examples? `Make style`
> * Should we add a small page in docs?

Thanks! These suggestions make sense to me. Will try to address later in the week!

TristanThrush commented:
Alright @lvwerra thanks again for your useful comments! In response, I:

* used AutoModelForSequenceClassification instead
* used a custom trainer with a custom `compute_loss` function
* I tried to not use a custom data collator, but I found that it was difficult for 2 reasons:
    1) I needed to specify `"return_loss": True` somewhere in the batched input data. If I don't do this, then `AutoModelForSequenceClassification` will error out, because my custom `compute_loss` function will not be used for the validation step. The custom `compute_loss` function needs to be used for training and validation instead of the normal model `forward` function. The normal `forward` function expects `input_ids`, not `input_ids_j` and `input_ids_k`. It is fine to use the normal `forward` function after the reward model is trained when you want to use it in the ppo stage, though.
    2) It seems a bit slower to pad to the max and then remove padding
    
I'm happy to try to make it work though if you still think that a custom data collator is not good!

TristanThrush commented:
Thanks @natolambert for the comments. In response, I linted my code and added the doc pages. The one thing I havent addressed is pointing to trained model which matches the paper's results. I will do so soon, but am still training a model on the newly altered code.

lvwerra commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,198 @@
+def turn_into_text_classification_format(examples):
+    new_examples = {"text_j": [], "text_k": []}
+    for info, summaries, choice in zip(examples["info"], examples["summaries"], examples["choice"]):
+        assert len(summaries) == 2
+        assert choice in (0, 1)
```
Can we do proper errors with useful messages? :)
```suggestion
        if len(summaries) != 2 or choice not in (0, 1):
            raise ValueError(f"There should be two summaries with a choice that's either 0 or 1. Received {len(summaries} summaries and choice={choice}.")
```

lvwerra commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,198 @@
+    learning_rate=script_args.learning_rate,
+    per_device_train_batch_size=script_args.per_device_train_batch_size,
+    per_device_eval_batch_size=script_args.per_device_eval_batch_size,
+    num_train_epochs=5,
```
That seems like a parameter someone might also want to tune, right? so maybe move it to the arguments as well.

TristanThrush replied:
Yeah really good point. 5 is the number of epochs that openai used, but it makes sense for me to make it an option and just make 5 the default. And then put that in the documentation.

lvwerra commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,198 @@
+    per_device_train_batch_size=script_args.per_device_train_batch_size,
+    per_device_eval_batch_size=script_args.per_device_eval_batch_size,
+    num_train_epochs=5,
+    weight_decay=0.001,
```
same here probably

lvwerra commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,198 @@
+    return accuracy.compute(predictions=predictions, references=labels)
+
+
+class RewardTrainer(Trainer):
```
Maybe at some point we can a general `RewardTrainer` to TRL which could also be used to train on the ranking task!

lvwerra commented on examples/summarization/scripts/reward_summarization.py:
```diff
@@ -0,0 +1,198 @@
+
+
+num_proc = (
+    100
```
it looks like a very quick operation so going for 100 processes might actually be more overhead than if we went maybe for 8?

TristanThrush replied:
yeah good point. The dataset size is fixed, after all.

lvwerra reviewed (commented):
Awesome, thanks for the refactoring. Looks much cleaner now IMO!

Left just a few minor things to clean up then we can merge! 🚀

natolambert commented:
Random comment on reward model, make sure we aren't assigning reward / gradients to the prompt / instruction tokens.

lvwerra commented:
@natolambert you mean during RL training? The PPOTrainer takes care of that and only looks at rewards for the response tokens. For the RM training it doesn't really matter as we predict the reward after the last token, right?

TristanThrush commented:
> Awesome, thanks for the refactoring. Looks much cleaner now IMO!
> 
> Left just a few minor things to clean up then we can merge! 🚀

Ok I believe I've addressed the comments. Will also add a model card to the linked model. Thanks, and let me know if there is anything else!
